src/stan/mcmc/hmc/nuts_classic/base_nuts_classic.hpp

```diff
@@ -0,0 +1,249 @@
+#ifndef STAN_MCMC_HMC_NUTS_BASE_NUTS_HPP
+#define STAN_MCMC_HMC_NUTS_BASE_NUTS_HPP
+
+#include <stan/interface_callbacks/writer/base_writer.hpp>
+#include <boost/math/special_functions/fpclassify.hpp>
+#include <stan/mcmc/hmc/base_hmc.hpp>
+#include <stan/mcmc/hmc/hamiltonians/ps_point.hpp>
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace mcmc {
+
+    struct nuts_util {
+      // Constants through each recursion
+      double log_u;
+      double H0;
+      int sign;
+
+      // Aggregators through each recursion
+      int n_tree;
+      double sum_prob;
+      bool criterion;
+    };
+
+    // The No-U-Turn Sampler (NUTS) with the
+    // original slice sampler implementation
+    template <class Model, template<class, class> class Hamiltonian,
+              template<class> class Integrator, class BaseRNG>
+    class base_nuts_classic:
+      public base_hmc<Model, Hamiltonian, Integrator, BaseRNG> {
+    public:
+      base_nuts_classic(Model &model, BaseRNG& rng):
+        base_hmc<Model, Hamiltonian, Integrator, BaseRNG>(model, rng),
+        depth_(0), max_depth_(5), max_delta_(1000),
+        n_leapfrog_(0), divergent_(0), energy_(0) {
+      }
+
+      ~base_nuts_classic() {}
+
+      void set_max_depth(int d) {
+        if (d > 0)
+          max_depth_ = d;
+      }
+
+      void set_max_delta(double d) {
+        max_delta_ = d;
+      }
+
+      int get_max_depth() { return this->max_depth_; }
+      double get_max_delta() { return this->max_delta_; }
+
+      sample transition(sample& init_sample,
+                        interface_callbacks::writer::base_writer& writer) {
+        // Initialize the algorithm
+        this->sample_stepsize();
+
+        nuts_util util;
+
+        this->seed(init_sample.cont_params());
+
+        this->hamiltonian_.sample_p(this->z_, this->rand_int_);
+        this->hamiltonian_.init(this->z_, writer);
+
+        ps_point z_plus(this->z_);
+        ps_point z_minus(z_plus);
+
+        ps_point z_sample(z_plus);
+        ps_point z_propose(z_plus);
+
+        int n_cont = init_sample.cont_params().size();
+
+        Eigen::VectorXd rho_init = this->z_.p;
+        Eigen::VectorXd rho_plus(n_cont); rho_plus.setZero();
+        Eigen::VectorXd rho_minus(n_cont); rho_minus.setZero();
+
+        util.H0 = this->hamiltonian_.H(this->z_);
+
+        // Sample the slice variable
+        util.log_u = std::log(this->rand_uniform_());
+
+        // Build a balanced binary tree until the NUTS criterion fails
+        util.criterion = true;
+        int n_valid = 0;
+
+        this->depth_ = 0;
+        this->divergent_ = 0;
+
+        util.n_tree = 0;
+        util.sum_prob = 0;
+
```

Comment:
I'd prefer to see util constructed with all its pieces in one line, but I can see how you'd like the names here.